src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java

```diff
@@ -0,0 +1,1753 @@
+import java.sql.SQLXML;
+import java.sql.Statement;
+import java.sql.Timestamp;
+import java.sql.Types;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.confluent.connect.jdbc.dialect.DatabaseDialectProvider.FixedScoreProvider;
+import io.confluent.connect.jdbc.sink.JdbcSinkConfig;
+import io.confluent.connect.jdbc.sink.JdbcSinkConfig.InsertMode;
+import io.confluent.connect.jdbc.sink.JdbcSinkConfig.PrimaryKeyMode;
+import io.confluent.connect.jdbc.sink.PreparedStatementBinder;
+import io.confluent.connect.jdbc.sink.metadata.FieldsMetadata;
+import io.confluent.connect.jdbc.sink.metadata.SchemaPair;
+import io.confluent.connect.jdbc.sink.metadata.SinkRecordField;
+import io.confluent.connect.jdbc.source.ColumnMapping;
+import io.confluent.connect.jdbc.source.JdbcSourceConnectorConfig;
+import io.confluent.connect.jdbc.source.JdbcSourceTaskConfig;
+import io.confluent.connect.jdbc.source.TimestampIncrementingCriteria;
+import io.confluent.connect.jdbc.util.ColumnDefinition;
+import io.confluent.connect.jdbc.util.ColumnDefinition.Mutability;
+import io.confluent.connect.jdbc.util.ColumnDefinition.Nullability;
+import io.confluent.connect.jdbc.util.ColumnId;
+import io.confluent.connect.jdbc.util.DateTimeUtils;
+import io.confluent.connect.jdbc.util.ExpressionBuilder;
+import io.confluent.connect.jdbc.util.ExpressionBuilder.Transform;
+import io.confluent.connect.jdbc.util.IdentifierRules;
+import io.confluent.connect.jdbc.util.JdbcDriverInfo;
+import io.confluent.connect.jdbc.util.TableDefinition;
+import io.confluent.connect.jdbc.util.TableId;
+
+/**
+ * A {@link DatabaseDialect} implementation that provides functionality based upon JDBC and SQL.
+ *
+ * <p>This class is designed to be extended as required to customize the behavior for a specific
+ * DBMS. For example, override the {@link #createColumnConverter(ColumnMapping)} method to customize
+ * how a column value is converted to a field value for use in a {@link Struct}. To also change the
+ * field's type or schema, also override the {@link #addFieldToSchema} method.
+ */
+public class GenericDatabaseDialect implements DatabaseDialect {
+
+  /**
+   * The provider for {@link GenericDatabaseDialect}.
+   */
+  public static class Provider extends FixedScoreProvider {
+    public Provider() {
+      super(GenericDatabaseDialect.class.getSimpleName(),
+            DatabaseDialectProvider.AVERAGE_MATCHING_SCORE
+      );
+    }
+
+    @Override
+    public DatabaseDialect create(AbstractConfig config) {
+      return new GenericDatabaseDialect(config);
+    }
+  }
```

Comment:
The `GenericDatabaseProvider` is one of the few of our dialect implementations that uses the `FixedScoreProvider` base class. The `GenericDatabaseProvider` is the _default_ provider, so it always returns the "average" score, which is lower than a "perfect" score (when the URL has a matching subprotocol supported by the dialect) but higher than a "no match" score (when the URL does not have a matching subprotocol).